src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -412,11 +412,16 @@ private function processAnonymousServices(\DOMDocument $xml, $file, $defaults)
 
         // anonymous services "in the wild"
         if (false !== $nodes = $xpath->query('//container:services/container:service[not(@id)]')) {
+            $num = 0;
             foreach ($nodes as $node) {
                 // give it a unique name
                 $id = sprintf('%d_%s', ++$count, hash('sha256', $file));
                 $node->setAttribute('id', $id);
                 $definitions[$id] = array($node, $file, true);
+                ++$num;
+            }
+            if ($num) {
+                @trigger_error(sprintf('Anonymous services (%d) in file %s are deprecated since 3.4 and wont be supported in 4.0', $num, $file), E_USER_DEPRECATED);
```

Comment:
imho, no need for $num. To be a bit more explicit, I propose:
Anonymous services are deprecated since Symfony 3.4, the "id" attribute will be required in version 4.0 in %s.